sdk/core/System.ClientModel/src/Internal/ArrayBackedPropertyBag.cs

```diff
@@ -0,0 +1,308 @@
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private bool IsFirst(TKey key) => _first.Key.Equals(key);
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private bool IsSecond(TKey key) => _second.Key.Equals(key);
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void AddInternal(TKey key, TValue value)
+    {
+        switch (_count)
+        {
+            case 0:
+                _first = new Kvp(key, value);
+                _count = 1;
+                return;
+            case 1:
+                if (IsFirst(key))
+                {
+                    _first = new Kvp(_first.Key, value);
+                }
+                else
+                {
+                    _second = new Kvp(key, value);
+                    _count = 2;
+                }
+
+                return;
+            default:
+                if (_rest == null)
+                {
+                    _rest = ArrayPool<Kvp>.Shared.Rent(8);
+                    _rest[_count++ - 2] = new Kvp(key, value);
+                    return;
+                }
+
+                if (_rest.Length <= _count)
+                {
+                    var larger = ArrayPool<Kvp>.Shared.Rent(_rest.Length << 1);
+                    _rest.CopyTo(larger, 0);
+                    var old = _rest;
+                    _rest = larger;
+                    ArrayPool<Kvp>.Shared.Return(old, true);
+                }
+                _rest[_count++ - 2] = new Kvp(key, value);
+                return;
+        }
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void SetAt(int index, Kvp value)
+    {
+        if (index == 0)
+            _first = value;
+        else if (index == 1)
+            _second = value;
+        else
+            GetRest()[index - 2] = value;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private TValue GetAt(int index) => index switch
+    {
+        0 => _first.Value,
+        1 => _second.Value,
+        _ => GetRest()[index - 2].Value
+    };
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private int GetIndex(TKey key)
+    {
+        if (_count == 0)
+            return -1;
+        if (_count > 0 && _first.Key.Equals(key))
+            return 0;
+        if (_count > 1 && _second.Key.Equals(key))
+            return 1;
+
+        if (_count <= 2)
+            return -1;
+
+        Kvp[] rest = GetRest();
+        int max = _count - 2;
+        for (var i = 0; i < max; i++)
+        {
+            if (rest[i].Key.Equals(key))
+                return i + 2;
+        }
+        return -1;
+    }
+
+    public void Dispose()
+    {
+#if DEBUG
+        if (_disposed)
+        {
+            return;
+        }
+        _disposed = true;
+#endif
+        _count = 0;
+        _first = default;
+        _second = default;
+        if (_rest == default)
+        {
+            return;
+        }
+
+        var rest = _rest;
```

Comment:
this is not thread safe and so there is a possibility of double return to the pool. Which is very severe; more severe than use -after-free. And so I think we should wrap these three lines of code into a lock.

Reply:
Could we not instead mark this field as `volatile`? Alternatively, if Chris's comment was addressed and we removed the DEBUG ifdef, we could mark `_disposed` as `volatile` and avoid locking.

Reply:
volatile is not enough. All these lines of code need to be synchronized.

Reply:
Addressed with a lock

Reply:
Will this work?

```cs
var rest = Interlocked.Exchange(ref _rest, default);
if (rest != default)
{
    ArrayPool.Shared.Return(rest, true);
}
```

Reply:
@AlexanderSher, is what I just pushed ok?  We can follow up with update PRs to main for improvements if needed.